quinn-proto/src/connection/assembler.rs

```diff
@@ -98,34 +108,106 @@ impl Assembler {
 
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
+        self.buffered = fragmented_buffered;
         for chunk in old.into_sorted_vec().into_iter().rev() {
-            let end = offset + (buffer.len() as u64);
-            if let Some(overlap) = end.checked_sub(chunk.offset) {
-                if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
-                    buffer.extend_from_slice(bytes);
-                }
-            } else {
-                let bytes = buffer.split().freeze();
-                self.data.push(Buffer { offset, bytes });
+            if chunk.is_defragmented() {
+                self.buffered += chunk.size;
+                self.data.push(chunk);
+                continue;
+            } else if !chunk.should_defragment() {
+                self.buffered += chunk.bytes.len();
+                self.data
+                    .push(Buffer::new_defragmented(chunk.offset, chunk.bytes));
+                continue;
+            }
+            if chunk.offset != offset + (buffer.len() as u64) {
+                self.data
+                    .push(Buffer::new_defragmented(offset, buffer.split().freeze()));
                 offset = chunk.offset;
-                buffer.extend_from_slice(&chunk.bytes);
             }
+            buffer.extend_from_slice(&chunk.bytes);
         }
+        self.data
+            .push(Buffer::new_defragmented(offset, buffer.split().freeze()));
+        self.allocated = self.buffered;
+    }
 
-        let bytes = buffer.split().freeze();
-        self.data.push(Buffer { offset, bytes });
-        self.defragmented = self.data.len();
+    fn deduplicate(&mut self) -> RangeSet {
+        debug_assert!(self.state.is_ordered());
+        let mut ranges = RangeSet::new();
+        ranges.insert(0..self.bytes_read);
+        let new = BinaryHeap::with_capacity(self.data.len());
+        let old = mem::replace(&mut self.data, new).into_vec();
+        for buffer in old.iter() {
```

Comment:
`defragment` reverse iterates. This forward iterates. Why the difference?

Reply:
In `defragment`, we want to iterate with ascending `offset` so that we can coalesce multiple `Buffer` into one. This is not required for `deduplicate`. Also, iterating it in the order of the underlying `Vec` makes sure that `self.data.push` is optimal (if we do not find any duplicates).

Reply:
This is also now obsolete, because I removed `deduplicate`.